File: make-TAPE5.py
import numpy as np
import math
import subprocess
import shutil

lev = 55
prof = 50
lev1 = 55

p = np.zeros((lev, prof), dtype=np.float32)
t = np.zeros((lev, prof), dtype=np.float32)
ozo = np.zeros((lev, prof), dtype=np.float32)
wv = np.zeros((lev, prof), dtype=np.float32)

p1 = np.zeros((lev1, prof), dtype=np.float32)
t1 = np.zeros((lev1, prof), dtype=np.float32)
ozo1 = np.zeros((lev1, prof), dtype=np.float32)
wv1 = np.zeros((lev1, prof), dtype=np.float32)

tem = np.zeros((prof, lev), dtype=np.float32)
H2O = np.zeros((prof, lev), dtype=np.float32)
o3 = np.zeros((prof, lev), dtype=np.float32)

tra2 = np.zeros((lev1, 4), dtype=np.float32)
trah2 = np.zeros((lev1, 4), dtype=np.float32)
tra = np.zeros((lev, 4), dtype=np.float32)
trah = np.zeros((lev, 4), dtype=np.float32)

angle = 0.0
va = np.zeros(6, dtype=np.float32)
height = np.zeros(lev, dtype=np.float32)
pp = np.zeros(lev, dtype=np.float32)
tsurf = np.zeros(prof, dtype=np.float32)
scant = np.array([2.58, 2.94, 3.72, 4.83, 6.1, 7.2, 9], dtype=np.float32)
blank = np.zeros(lev, dtype=np.float32)
bwn = 0.0
ewn = 0.0
maxwn = 1000.0
bwn0 = 6500.0
ewn0 = 8500.0
lay = ['001', '002', '003', '004', '005', '006', '007', '008', '009', '010',
       '011', '012', '013', '014', '015', '016', '017', '018', '019', '020',
       '021', '022', '023', '024', '025', '026', '027', '028', '029', '030',
       '031', '032', '033', '034', '035', '036', '037', '038', '039', '040',
       '041', '042', '043', '044', '045', '046', '047', '048', '049', '050',
       '051', '052', '053', '054']

# Read pressure data
with open('./profile/pressure.txt', 'r') as file:
    for i in range(lev):
        p1[i, :] = np.array(file.readline().split(), dtype=np.float32)
for i in range(lev):
    p[i, :] = p1[lev - i - 1, :]

p[p < 0.001] = 0.001

# Read temperature data
with open('./profile/temperature.txt', 'r') as file:
    for i in range(lev):
        t1[i, :] = np.array(file.readline().split(), dtype=np.float32)
for i in range(lev):
    t[i, :] = t1[lev - i - 1, :]

# Read ozone data
# with open('./profile/ozo_test_1573.txt', 'r') as file:
#     for i in range(lev):
#         ozo1[i, :] = np.array(file.readline().split(), dtype=np.float32)
# for i in range(lev):
#     ozo[i, :] = ozo1[lev - i - 1, :]

# Read water vapor data
with open('./profile/wv.txt', 'r') as file:
    for i in range(lev):
        wv1[i, :] = np.array(file.readline().split(), dtype=np.float32)
for i in range(lev):
    wv[i, :] = wv1[lev - i - 1, :]

for i in range(6):
    angle = 1 + 0.25 * i
    va[i] = math.degrees(math.acos(1 / angle))

height.fill(0.0)
blank.fill(0.0)

for iprof in range(prof):
    tsurf[iprof] = t[0, iprof]

for iadd in range(1):
    bwn = bwn0 + iadd * maxwn
    ewn = ewn0 + iadd * maxwn

    for jj in range(1):
        c1 = str(jj + 1)
        for iprof in range(prof):
            c2 = f'{iprof + 1:04d}'
            with open('TAPE5', 'w') as file:
                CXID = '83P 55level profile to creat transmittance database'
                file.write(f"${CXID}\n")
                file.write(f" HI=1 F4=1 CN=1 AE=0 EM=0 SC=0 FI=0 PL=0 TS=1 AM=0 MG=0 LA=0 OD=1 XS=0   00   00\n")
                file.write(f"{bwn:10.3f}{ewn:10.3f}                                                  REJ=0     1.000E-01\n")
                file.write(f"{0:7}{2:5}{-55:5}{1:5}{1:5}{7:5}{1:5}{0:5}{0:10.3f}{0:10.3f}\n")
                file.write(f"{p[lev-1, iprof]:10.3f}{p[0, iprof]:10.3f}{180.000:10.3f}\n")
                column_data = p[:, iprof]
                # 以8个一组分批写入数据
                file.writelines([f"{' '.join([f'{x:10.3f}' for x in column_data[i:i+8]])}\n" for i in range(0, len(column_data), 8)])
                file.write(f"{-55:5}{'INPUT FOR CAMEX':9}\n")
                for i in range(lev):
                    file.write(f"{height[i]:10.3f}{p[i, iprof]:10.3f}{t[i, iprof]:10.3f}  {'AA'} {'L'} {'AAAAAAA'}\n")
                    file.write(f"{wv[i, iprof]:15.8e}{blank[i]:15.8e}{blank[i]:15.8e}{blank[i]:15.8e}{blank[i]:15.8e}{blank[i]:15.8e}{blank[i]:15.8e}\n")
                file.write(f"-1.\n")
                file.write(f"$ Transfer to ASCII plotting data\n")
                file.write(f" HI=0 F4=0 CN=0 AE=0 EM=0 SC=0 FI=0 PL=1 TS=0 AM=0 MG=0 LA=0 OD=0 XS=0 0 0 0\n")
                file.write(f"# Plot title not used\n")
                file.write(f"{bwn:10.4f}{ewn:10.4f}{10.2000:10.4f}{100.000:10.4f}{5:5}{0:5}{12:5}{0:5}{1.000:10.3f}{0:5}{0:5}{0:5}\n")
                file.write(f"{0.0000:10.4f}{1.2000:10.4f}{7.0200:10.3f}{0.2000:10.3f}{4:5}{0:5}{1:5}{1:5}{0:5}{0:5}{0:5}{3:5}{27:5}\n")
                file.write(f"{bwn:10.4f}{ewn:10.4f}{10.2000:10.4f}{100.000:10.4f}{5:5}{0:5}{12:5}{0:5}{1.000:10.3f}{0:5}{0:5}{0:5}\n")
                file.write(f"{0.0000:10.4f}{1.2000:10.4f}{7.0200:10.3f}{0.2000:10.3f}{4:5}{0:5}{1:5}{1:5}{0:5}{0:5}{0:5}{3:5}{28:5}\n")
                file.write(f"-1.\n")
                file.write(f"%%%%%\n")
            # TAPE5制作好后，运行pl脚本，pl脚本中写着运行模型和输出提示的命令
            subprocess.run(['./script_run_example.pl'], check=True)
            bwn_str = f"{bwn:9.3f}".strip()
            source = 'ODint_001'
            destination = f'./OD001save_H2O/ODint_001-{bwn_str}'
            shutil.move(source, destination)

File: make-TAPE5.f90
program lbl_input
implicit none

integer   lev,prof,lev1
parameter (lev=55,prof=50,lev1=55)!1039
real*4    p(lev,prof),t(lev,prof),ozo(lev,prof),wv(lev,prof)
real*4    p1(lev1,prof),t1(lev1,prof),ozo1(lev1,prof),wv1(lev1,prof)
real*4    tem(prof,lev),H2O(prof,lev),o3(prof,lev)
real*4    tra2(lev1,4),trah2(lev1,4),tra(lev,4),trah(lev,4)
integer   i,j,iprof,jj,kk,iii,jjj,iadd
real*4    angle,va(6),height(lev),pp(lev),tsurf(prof),scant(7)
real*4    blank(lev),bwn,ewn
real*4    maxwn,bwn0,ewn0
character*3 lay(54)
CHARACTER CXID*52
character(len=20) :: bwn_str
character*4 :: c2
character*1 :: c1

data lay/'001','002','003','004','005','006','007','008','009','010','011','012','013','014','015','016','017','018','019','020',&
'021','022','023','024','025','026','027','028','029','030','031','032','033','034','035','036','037','038','039','040',&
'041','042','043','044','045','046','047','048','049','050','051','052','053','054'/!,'055','056','057','058','059','060',&
!'061','062','063','064','065','066','067','068','069','070','071','072','073','074','075','076','077','078','079','080',&
!'081','082','083','084','085','086','087','088','089','090','091','092','093','094','095','096','097','098','099','100'/
data scant/2.58,2.94,3.72,4.83,6.1,7.2,9/



open(11,file='./profile/pressure.txt',form='formatted')
do i=1,lev
   read(11,*) p1(i,:)
end do
close(11)
do i=1,lev
   p(i,:)=p1(lev+1-i,:)
end do



open(12,file='./profile/temperature.txt',form='formatted')
do i=1,lev
   read(12,*) t1(i,:)
end do
close(12)
do i=1,lev
   t(i,:)=t1(lev+1-i,:)
end do



open(13,file='./profile/ozo_test_1573.txt',form='formatted')
do i=1,lev
   read(13,*) ozo1(i,:)
end do
close(13)
do i=1,lev
   ozo(i,:)=ozo1(lev+1-i,:)
end do


open(14,file='./profile/wv.txt',form='formatted')
do i=1,lev
   read(14,*) wv1(i,:)
end do
close(14)
do i=1,lev
   wv(i,:)=wv1(lev+1-i,:)
end do




do i=1,6
angle=1+0.25*(i-1)
va(i)=acos(1./angle)
va(i)=va(i)*180./3.1415927
enddo 

  
height=0.00
blank=0.00




do iprof=1,prof
tsurf(iprof)=t(1,iprof)
end do

!%%%%% for fixed gas
maxwn=1000.000
bwn0=6500.000
ewn0=8500.000
do iadd=0,0
bwn=bwn0+iadd*maxwn
ewn=ewn0+iadd*maxwn
!bwn=14500.000
!ewn=16500.000
!bwn=16500.000
!ewn=18500.000
!bwn=18500.000
!ewn=20500.000
!bwn=20500.000
!ewn=22500.000
!bwn=22500.000
!ewn=24500.000
!bwn=24500.000
!ewn=26500.000
!bwn=26500.000
!ewn=28500.000
!bwn=28500.000
!ewn=30500.000


do jj=1,1!angle
write( c1,'(i1)' ) jj
do iprof=1,50!prof
write( c2,'(i4)' ) iprof
open(44,file='TAPE5',status='REPLACE')
CXID='83P 55level profile to creat transmittance database'
write(44,"(A1,A52)") '$',CXID
write(44,"(1X,'HI=',I1,1X,'F4=',I1,1X,'CN=',I1,1X,'AE=',I1,1X,'EM=',I1,1X,'SC=',I1,1X,'FI=',I1,1X,'PL=',I1,1X,'TS=',I1,1X,'AM=',I1,1X,'MG=',I1,1X,'LA=',I1,1X,'OD=',I1,1X,'XS=',I1,3X,I1,I1,3X,I1,I1)") 1,1,1,0,0,0,0,0,0,1,0,0,1,0,0,0,0,0!lbl的设置参数，笔记上有
write(44,"(2F10.3,60X,'REJ=',I1,5X,E10.3)") bwn,ewn,0,0.1!开始、结束波数和分辨率
!write(44,"(2F10.3,20X,F10.3)") tsurf(iprof),1.0000,0.0000！地表参数（经常出问题）
write(44,"(7I5,I2,1X,I2,2F10.3)")0,2,-55,1,1,7,1,0,0,0.0000,0.0000!层数、气体数量，最后两个是透过率和辐射率（？）
write(44,"(3F10.3)") p(55,iprof),p(1,iprof),180.000!0.00,1009.00,180.000!-va(jj)！最大最小气压值，和角度，180就是天顶角
write(44,"(8F10.3)") p(:,iprof)
write(44,"(I5,9X,3A8)")-55,'INPUT FOR CAMEX'!这里跟上面的两个层数要一致
do i=1,lev
WRITE(44,"(3F10.3,5X,A2,1X,A1,1X,7A)") height(i),p(i,iprof),t(i,iprof),'AA','L','AAAAAAA'!'AAAAAAAAAAAAAAAAAAAAAA'!这里开几个气体写几个A
!WRITE(44,"(8E15.8)") wv(i,iprof),blank(i),ozo(i,iprof),blank(i),blank(i),blank(i),blank(i),blank(i),blank(i),blank(i),blank(i),blank(i),blank(i),blank(i),blank(i),blank(i),blank(i),blank(i),blank(i),blank(i),blank(i),blank(i)
WRITE(44,"(8E15.8)") wv(i,iprof),blank(i),blank(i),blank(i),blank(i),blank(i),blank(i)
!WRITE(44,"(8E15.8)") wv(i,iprof),tra(i,1),ozo(i,iprof),blank(i),blank(i),blank(i),blank(i)!,blank(i),blank(i),blank(i),blank(i),blank(i),tra(i,10),blank(i),blank(i),tra(i,11)!这里开几个气体跟上面的A保持一致，不用的气体就blank
end do

!惰性气体
!WRITE(44,"(3I5)") 4,0,0
!WRITE(44,"(4A10)") 'CCL4','CFC11','CFC12','CFC14'  
!WRITE(44,"(I5,I5,A50)") 101,1,'CAMEX Wallops'
!do i=1,lev
!write(44,"(F10.3,5X,A4)") p(i),'AAAA'
!write(44,"(4E10.3)") trah(i,:)
!write(44,"(4E10.3)") blank(i),blank(i),blank(i),blank(i)
!write(44,"(4E10.3)") trah(i,:)
!end do

!这一部分不用管
write(44,"(F4.1)") -1.
                  
write(44,"(A1,1X,A30)") '$','Transfer to ASCII plotting data'
write(44,"(1X,'HI=',I1,1X,'F4=',I1,1X,'CN=',I1,1X,'AE=',I1,1X,'EM=',I1,1X,'SC=',I1,1X,'FI=',I1,1X,'PL=',I1,1X,'TS=',I1,1X,'AM=',I1,1X,'MG=',I1,1X,'LA=',I1,1X,'OD=',I1,1X,'XS=',I1,3X,I1,3X,I1)") 0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0
WRITE(44,"(A1,1X,19A)")'#','Plot title not used'
write(44,"(4F10.4,4I5,F10.3,I2,I3,I5)") bwn,ewn,10.2000,100.000,5,0,12,0,1.0000,0,0,0
write(44,"(2F10.4,2F10.3,6I5,I2,3X,I2,I3)") 0.0000,1.2000,7.0200,0.2000,4,0,1,1,0,0,0,3,27
write(44,"(4F10.4,4I5,F10.3,I2,I3,I5)") bwn,ewn,10.2000,100.000,5,0,12,0,1.0000,0,0,0
write(44,"(2F10.4,2F10.3,6I5,I2,3X,I2,I3)") 0.0000,1.2000,7.0200,0.2000,4,0,1,1,0,0,0,3,28
write(44,"(F4.1)") -1.
write(44,"(A5)") '%%%%%'


! call system('./script_run_example.pl')
! write(bwn_str, '(F9.3)') bwn
! call system('mv ODint_001 ./OD001save_H2O/ODint_001-'//trim(adjustl(bwn_str)))

!call system('mv TAPE12_test test_'//trim(adjustl(c2))//'P_angel'//trim(adjustl(c1))//'_dt ')
!call system('mv test_'//trim(adjustl(c2))//'P_angel'//trim(adjustl(c1))//'_dt ./lblrtm_transmittance_both_1/')!tape12
!do kk=1,54!lay
!call system('mv ODint_'//lay(kk)//' test_'//trim(adjustl(c2))//'P_angel'//trim(adjustl(c1))//'_od_'//lay(kk)//'')
!call system('mv test_'//trim(adjustl(c2))//'P_angel'//trim(adjustl(c1))//'_od_'//lay(kk)//' ./lblrtm_od_and_others/')!001第一层(layer)，1P第一个profile，od光学厚度，即tape11

!end do
close(44)
end do
end do
end do

end program lbl_input
